src/DDTrace/Integrations/Memcached.php

```diff
@@ -0,0 +1,358 @@
+        $span->setTag(Tags\SPAN_TYPE, Types\MEMCACHED);
+        $span->setTag(Tags\SERVICE_NAME, 'memcached');
+        $span->setTag('memcached.command', 'casByKey');
+        $span->setTag('memcached.cas_token', $args[0]);
+
+        $serverKey = $args[1];
+        $key = $args[2];
+        $span->setTag('memcached.server_key', $serverKey);
+        $span->setTag('memcached.query', "casByKey $key");
+        $span->setResource('casByKey');
+        self::setServerTagsByKey($span, $memcached, $serverKey);
+
+        try {
+            return $memcached->casByKey(...$args);
+        } catch (\Exception $e) {
+            $span->setError($e);
+            throw $e;
+        } finally {
+            $scope->close();
+        }
+    }
+
+    public static function traceMulti($memcached, $command, $args)
+    {
+        $scope = GlobalTracer::get()->startActiveSpan("Memcached.$command");
+        $span = $scope->getSpan();
+        $span->setTag(Tags\SPAN_TYPE, Types\MEMCACHED);
+        $span->setTag(Tags\SERVICE_NAME, 'memcached');
+        $span->setTag('memcached.command', $command);
+
+        $query = "$command " . implode(',', array_keys($args[0]));
+        $span->setTag('memcached.query', $query);
+        $span->setResource($command);
+
+        try {
+            return $memcached->$command(...$args);
+        } catch (\Exception $e) {
+            $span->setError($e);
+            throw $e;
+        } finally {
+            $scope->close();
+        }
+    }
+
+    public static function traceMultiByKey($memcached, $command, $args)
+    {
+        $scope = GlobalTracer::get()->startActiveSpan("Memcached.$command");
+        $span = $scope->getSpan();
+        $span->setTag(Tags\SPAN_TYPE, Types\MEMCACHED);
+        $span->setTag(Tags\SERVICE_NAME, 'memcached');
+        $span->setTag('memcached.command', $command);
+        $span->setTag('memcached.server_key', $args[0]);
+        self::setServerTagsByKey($span, $memcached, $args[0]);
+
+        $query = "$command " . implode(',', array_keys($args[1]));
+        $span->setTag('memcached.query', $query);
+        $span->setResource($command);
+
+        try {
+            return $memcached->$command(...$args);
+        } catch (\Exception $e) {
+            $span->setError($e);
+            throw $e;
+        } finally {
+            $scope->close();
+        }
+    }
+
+    /**
+     * Memcached::getServerByKey() /might/ return incorrect information if the
+     * distribution would be rebuilt on a real call (Memcached::get(),
+     * Memcached::getByKey(), and other commands that actually hit the server
+     * include logic to regenerate the distribution if a server has been ejected
+     * or if a timer expires; Memcached::getServerByKey() does not check for the
+     * distribution being rebuilt. Getting around that would likely be
+     * prohibitively expensive though.
+     */
+    public static function setServerTagsByKey($span, $memcached, $key)
```

Comment:
Should this be private or is intended to be used by devs?

Reply:
I think I had it public when it was called directly from a closure and missing class scope. All of the calls are from within the integration class now though, so it can be private. Changed.